soroban-env-host/src/vm.rs

```diff
@@ -260,8 +270,76 @@ impl Vm {
         VmInstantiationTimer::new(host.clone());
         let config = get_wasmi_config(host.as_budget())?;
         let engine = wasmi::Engine::new(&config);
-        let parsed_module = Rc::new(ParsedModule::new(host, &engine, wasm, cost_inputs)?);
-        Self::instantiate(host, contract_id, parsed_module)
+        let parsed_module = Self::parse_module(host, &engine, wasm, cost_inputs)?;
+        let linker = parsed_module.make_linker()?;
+        Self::instantiate(host, contract_id, parsed_module, &linker)
+    }
+
+    #[cfg(not(any(test, feature = "recording_mode")))]
+    fn parse_module(
+        host: &Host,
+        engine: &wasmi::Engine,
+        wasm: &[u8],
+        cost_inputs: VersionedContractCodeCostInputs,
+    ) -> Result<Rc<ParsedModule>, HostError> {
+        ParsedModule::new(host, engine, wasm, cost_inputs)
+    }
+
+    /// This method exists to support [crate::storage::FootprintMode::Recording]
+    /// when running in protocol versions that feature the [ModuleCache].
+    ///
+    /// There are two ways we can get to here:
+    ///
+    ///   1. When we're running in a protocol that doesn't support the
+    ///   [ModuleCache] at all. In this case, we just parse the module and
+    ///   charge for it as normal.
+    ///
+    ///   2. When we're in a protocol that _does_ support the [ModuleCache] but
+    ///   are _also_ in [crate::storage::FootprintMode::Recording] mode. Then
+    ///   the [ModuleCache] _does not get built_ during host setup (because we
+    ///   have no footprint yet to buid the cache from), so our caller
+    ///   [Host::call_contract_fn] sees no module cache, and so each call winds
+    ///   up calling us here, reparsing each module as it's called, and then
+    ///   throwing it away.
+    ///
+    /// When we are in case 2, we don't want to charge for all those reparses:
+    /// we want to charge only for the post-parse instantiations _as if_ we had
+    /// had the cache. The cache will actually be added in [Host::pop_context]
+    /// _after_ a top-level recording-mode invocation completes, by reading the
+    /// storage and parsing all the modules in it, in order to charge for
+    /// parsing each used module _once_ and thereby produce a mostly-correct
+    /// total cost.
+    ///
+    /// We still charge the reparses to the shadow budget, to avoid a DoS risk,
+    /// and we still charge the instantiations to the real budget, to behave the
+    /// same as if we had a cache.
+    ///
+    /// Note that this will only produce "mostly correct" costs for recording in
```

Comment:
Is this paragraph still true, given that we build the cache independently of the return value?

Reply:
No, you're right, it's misleading now. I'll remove that paragraph.